##  Week 3: Preparing for Exploring Data Analysis Using Python

*  Student Name: Abraham Abate
*  Instructor: Cary Jim
*  DSC 530: Data Exploration and Analysis
*  Data Science Dept., BU
*  Date: 06/22/2024

## Excerise 1.1:

In [12]:
from os.path import basename, exists
import numpy as np

def download(url):
    filename = basename(url)
    if not exists(filename):
        from urllib.request import urlretrieve

        local, _ = urlretrieve(url, filename)
        print("Downloaded " + local)


download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/thinkstats2.py")
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/thinkplot.py")
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/nsfg.py")
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/2002FemPreg.dct")
download("https://github.com/AllenDowney/ThinkStats2/raw/master/code/2002FemPreg.dat.gz")

In [13]:
# Read NSFG data into a Pandas DataFrame.
import nsfg

preg_df = nsfg.ReadFemPreg()
preg_df.head()

caseid  pregordr  howpreg_n  howpreg_p  moscurrp  nowprgdk  pregend1  \
0       1         1        NaN        NaN       NaN       NaN       6.0   
1       1         2        NaN        NaN       NaN       NaN       6.0   
2       2         1        NaN        NaN       NaN       NaN       5.0   
3       2         2        NaN        NaN       NaN       NaN       6.0   
4       2         3        NaN        NaN       NaN       NaN       6.0   

   pregend2  nbrnaliv  multbrth  ...  laborfor_i  religion_i  metro_i  \
0       NaN       1.0       NaN  ...           0           0        0   
1       NaN       1.0       NaN  ...           0           0        0   
2       NaN       3.0       5.0  ...           0           0        0   
3       NaN       1.0       NaN  ...           0           0        0   
4       NaN       1.0       NaN  ...           0           0        0   

       basewgt  adj_mod_basewgt      finalwgt  secu_p  sest  cmintvw  \
0  3410.389399      3869.349602   6448.271112       2     9      NaN   
1  3410.389399      3869.349602   6448.271112       2     9      NaN   
2  7226.301740      8567.549110  12999.542264       2    12      NaN   
3  7226.301740      8567.549110  12999.542264       2    12      NaN   
4  7226.301740      8567.549110  12999.542264       2    12      NaN   

   totalwgt_lb  
0       8.8125  
1       7.8750  
2       9.1250  
3       7.0000  
4       6.1875  

[5 rows x 244 columns]

In [14]:
# Print the column names.
# attribute columns returns a sequence of column names
preg_df.columns

Index(['caseid', 'pregordr', 'howpreg_n', 'howpreg_p', 'moscurrp', 'nowprgdk',
       'pregend1', 'pregend2', 'nbrnaliv', 'multbrth',
       ...
       'laborfor_i', 'religion_i', 'metro_i', 'basewgt', 'adj_mod_basewgt',
       'finalwgt', 'secu_p', 'sest', 'cmintvw', 'totalwgt_lb'],
      dtype='object', length=244)

In [15]:
# Select the birthord column, print the value counts
preg_df.birthord

0        1.0
1        2.0
2        1.0
3        2.0
4        3.0
        ... 
13588    1.0
13589    NaN
13590    NaN
13591    2.0
13592    3.0
Name: birthord, Length: 13593, dtype: float64

In [16]:
# use isnull to count the number of nans.
preg_df.birthord.isnull().sum()

4445

In [17]:
# Select a column and print the column
pregordr = preg_df['pregordr']
pregordr

0        1
1        2
2        1
3        2
4        3
        ..
13588    1
13589    2
13590    3
13591    4
13592    5
Name: pregordr, Length: 13593, dtype: int64

In [18]:
# Count the number of times each value occurs.
preg_df.outcome.value_counts().sort_index()

1    9148
2    1862
3     120
4    1921
5     190
6     352
Name: outcome, dtype: int64

In [19]:
# Select the prglngth column, print the value counts
preg_df.prglngth

0        39
1        39
2        39
3        39
4        39
         ..
13588    39
13589     6
13590     5
13591    39
13592    39
Name: prglngth, Length: 13593, dtype: int64

In [20]:
# To compute the mean of a column, you can invoke the mean method on a Series.
preg_df.totalwgt_lb.mean()

7.265628457623368

In [21]:
# Create a new column named totalwgt_kg that contains birth weight in kilograms.
preg_df["totalwgt_kg"] = preg_df["totalwgt_lb"] * 2.2
preg_df["totalwgt_kg"]

0        19.3875
1        17.3250
2        20.0750
3        15.4000
4        13.6125
          ...   
13588    13.6125
13589        NaN
13590        NaN
13591    16.5000
13592    16.5000
Name: totalwgt_kg, Length: 13593, dtype: float64

## Excerise 1.2:
* Write code that reads the respondent file, 2002FemResp.dat.gz. You might want to start with a copy of nsfg.py and modify it.
* The variable pregnum is a recode that indicates how many times each respondent has been pregnant. Print the value counts for this variable and compare them to the published results in the NSFG codebook.
* You can also cross-validate the respondent and pregnancy files by comparing pregnum for each respondent with the number of records in the pregnancy file.
* You can use nsfg.MakePregMap to make a dictionary that maps from each caseid to a list of indices into the pregnancy DataFrame



Select a column and check what type it is.

In [22]:
# access a column from a DataFrame
preg_num = preg_df['pregnum']
preg_num

0        2
1        2
2        3
3        3
4        3
        ..
13588    5
13589    5
13590    5
13591    5
13592    5
Name: pregnum, Length: 13593, dtype: int64

In [23]:
# check data type
type(preg_num)

pandas.core.series.Series

In [24]:
# count the number of times each respondent has been pregnant.
pregnum_count = preg_df.pregnum.value_counts().sort_index()
pregnum_count

1     1267
2     2864
3     3330
4     2444
5     1525
6      900
7      560
8      320
9      189
10      90
11      33
12      24
14      28
19      19
Name: pregnum, dtype: int64

Make a dictionary that maps from each respondent's `caseid` to a list of indices into the pregnancy `DataFrame`.  Use it to select the pregnancy number for a single respondent.

In [25]:
# Get the number of caseids using value count and convert it to a list
preg_count = preg_df['caseid'].value_counts().tolist()

# make a dictionary that maps from each caseid to a list of indices
preg_ct_dict = {}

for x in preg_count:
    preg_ct_dict[x] = preg_ct_dict.get(x, 0) + 1

preg_ct_dict

{19: 1,
 14: 2,
 12: 2,
 11: 3,
 10: 9,
 9: 21,
 8: 40,
 7: 80,
 6: 150,
 5: 305,
 4: 611,
 3: 1110,
 2: 1432,
 1: 1267}

In [26]:
# Sort the dictionary in descending order
sorted_lst = sorted(preg_ct_dict.items())
sorted_lst

[(1, 1267),
 (2, 1432),
 (3, 1110),
 (4, 611),
 (5, 305),
 (6, 150),
 (7, 80),
 (8, 40),
 (9, 21),
 (10, 9),
 (11, 3),
 (12, 2),
 (14, 2),
 (19, 1)]

In [27]:
# Average pregnancy length
avg_preg_len = preg_df.prglngth.mean()
avg_preg_len

29.531229309203265

In [28]:
# summary statistics for pregnancy number attribute with one baby
one_pregnum = preg_df[preg_df["pregnum"] == 1]
one_pregnum["pregnum"].describe()

count    1267.0
mean        1.0
std         0.0
min         1.0
25%         1.0
50%         1.0
75%         1.0
max         1.0
Name: pregnum, dtype: float64

In [29]:
# Mean pregnancy length for for first babies
first_babies = preg_df[preg_df["pregnum"] == 1]
first_babies["prglngth"].mean()

29.82004735595896

In [30]:
# Mean pregnancy length for for second babies
sec_babies = preg_df[preg_df["pregnum"] == 2]
sec_babies["prglngth"].mean()

32.20635474860335

In [31]:
# Mean pregnancy length for third, fourth,..babies
other_babies = preg_df[preg_df["pregnum"] == 3]
other_babies["prglngth"].mean()

30.80750750750751

In [32]:
# count how many babies born earlier than 40 weeks
longer_preg_len = preg_df[(preg_df["prglngth"] < 40) & (preg_df["pregnum"] == 1)]
longer_preg_len["prglngth"].describe()

count    1038.000000
mean       27.365125
std        13.938445
min         0.000000
25%        12.000000
50%        37.000000
75%        39.000000
max        39.000000
Name: prglngth, dtype: float64

In [33]:
# count how many babies born in 40 weeks
normal_preg_len = preg_df[(preg_df["prglngth"]== 40) & (preg_df["pregnum"] == 1)]
normal_preg_len["prglngth"].describe()

count    99.0
mean     40.0
std       0.0
min      40.0
25%      40.0
50%      40.0
75%      40.0
max      40.0
Name: prglngth, dtype: float64

In [34]:
# count how many babies born longer than 40 weeks
longer_preg_len = preg_df[(preg_df["prglngth"] > 40) & (preg_df["pregnum"] == 1)]
longer_preg_len["prglngth"].describe()

count    130.000000
mean      41.669231
std        0.959558
min       41.000000
25%       41.000000
50%       41.000000
75%       42.000000
max       47.000000
Name: prglngth, dtype: float64

In [35]:
# percentage of  first babies born in early, at 40 weeks and late
percent_early = (1038/1267) *100
print("Percentage for early born first babies: ", percent_early)
percent_normal = (99/1267) *100
print("Percentage for normal birth first babies: ", percent_normal)
percent_late = (130/1267) *100
print("Percentage for late born first babies: ", percent_late)

Percentage for early born first babies:  81.9258089976322
Percentage for normal birth first babies:  7.813733228097869
Percentage for late born first babies:  10.260457774269929


## Exercise 2-1

Based on the results in this chapter, suppose you were asked to summarize what you learned about whether first babies arrive late.


###### Which summary statistics would you use if you wanted to get a story on the evening news?
I would utlize the mean, median and standard deviation. The mean and median will help us for describing the location (or typical value) of a set of measurements/data points mainly where the center is located. Variation always exists among pregrnancies, regardless of which characteristic you’re measuring, because not every individual pregnancies will have the same exact outcome for every characteristic you measure. So, for that matter, to measure the variablity we can use Standard deviation. 

###### Which ones would you use if you wanted to reassure an anxious patient?
I would use the mean and Standard deviation here to show a comparative figure among different preganancies. This is to put the difference among different pregnancies in context.

###### “Do first babies arrive late?” Write a paragraph that uses the results in this chapter to answer the question clearly, precisely, and honestly.
While it’s true that only 7.8% of babies are born on their due dates, the idea that first babies are usually several days or even a week late isn’t entirely accurate. Looking at the average pregnancy length, first babies are born on 29.8 weeks and subsequent babies are born on 32.2 weeks. They’re more likely to be born early. We can validate that from the data analysis result, 81.9% of the first babies born earlier than 40 weeks and 10.3% of them born later than 40 weeks. Based on this analysis output, I would summize that the perception that first babies are usually late is incorrect and are in fact they were born earlier than other babies.

## Exercise 2-4.
* Using the variable totalwgt_lb, investigate whether first babies are lighter or heavier than others. Compute Cohen's d to quantify the difference between the groups. How does it compare to the difference in pregnancy length?

In [36]:
# Mean weight of first babies
avg_wt1 = preg_df[preg_df["pregnum"] == 1]
avg_wt1["totalwgt_lb"].mean()

7.173504146919432

In [37]:
# Mean weight of other babies
avg_wt2 = preg_df[preg_df["pregnum"] > 1]
avg_wt2["totalwgt_lb"].mean()

7.275117463998047

In [38]:
# Compute Cohen's d to quantify the difference between the groups.
def CohenEffectSize(group1, group2):
    """Computes Cohen's effect size for two groups.

    group1: Series or DataFrame
    group2: Series or DataFrame

    returns: float if the arguments are Series;
             Series if the arguments are DataFrames
    """
    diff = group1.mean() - group2.mean()

    var1 = group1.var()
    var2 = group2.var()
    n1, n2 = len(group1), len(group2)

    pooled_var = (n1 * var1 + n2 * var2) / (n1 + n2)
    d = diff / np.sqrt(pooled_var)
    return d

In [39]:
CohenEffectSize(avg_wt1["totalwgt_lb"], avg_wt2["totalwgt_lb"])

-0.07216530058667059

##### Interpretation:
* A Cohen’s d value of -0.072 indicates a small effect size. When interpreting Cohen’s d, negative values suggest that the treatment group (first babies) has a lower mean birth weight than the control group in this case other babies. However, in this case, the effect size is quite small, which means that the difference between the groups is not substantial. 